### DISCUSSION TITLE
Merge Queue Feedback

### DISCUSSION THREAD

**Discussion by benjaminmurphy** [community/community#14801] [Other Feature Feedback, Questions, & Ideas]
Like others, I can't find a place for feedback on the merge queue, so posting this here.

We tried enabling the new merge queue beta for our monorepo, and ended up turning it off after a couple hours. We like the idea in theory, but in practice, the implementation was missing some key features:

- Since every merge request has to re-run status checks separately before being added to the queue and before merging, this added a significant amount of time to our deploy process (an additional ~10-15 minutes per pull request, plus time waiting for merges ahead of you in the queue).
- No real introspection on the state of the queue. We ended up in the strange state where the pull request at the head of the queue had passed status checks and yet was waiting on the request _behind_ it, which feels like it should never happen.
- Insufficient parallelization. Testing two PRs at a time is great during low throughput hours of the day, but isn't enough during peak commit hours. This should either be tunable or learn from the length of the queue.
- Unclear interactions with GitHub Actions. We'd like to be able to define separate required status checks for adding to the merge queue and actually getting shipped from the merge queue to master. As an example, we have a code freeze action that checks for after-hours commits. We'd like that to run before adding a commit to the queue, but not on commits that are in the queue itself.

**Comment by danlucraft** [comment]
@benjaminmurphy 

> Insufficient parallelization. Testing two PRs at a time is great during low throughput hours of the day, but isn't enough during peak commit hours. This should either be tunable or learn from the length of the queue.

Totally agree on this. I'm looking at a queue with 8 PRs, that is going to take 2 hours to clear because our build takes 30 mins and the 2-build limit. If we could run 4 builds at a time it would clear in 1 hour. And as many as we like, 30 minutes!

**Comment by jgoux** [comment]
> Since every merge request has to re-run status checks separately before being added to the queue and before merging, this added a significant amount of time to our deploy process (an additional ~10-15 minutes per pull request, plus time waiting for merges ahead of you in the queue).

This is a major pain for us as well, it doubles the time we have to wait, the CI is running against the exact same code, it's unnecessary.
